### DISCUSSION TITLE
Question about output analysis of PF simulations

### DISCUSSION THREAD

**Discussion by abc-hy** [idaholab/moose#24720] [Q&A Modules: Phase field] [answered]
Hi everyone,

Do you know if MOOSE has any tools for analyzing the outputs of PF simulations? In particular, does it have a way of determining the characteristic dimensions of phase domains, e.g., using spectral analysis?

Thank you very much.
Best,

**Comment by GiudGiud** [comment]
@laagesen @amjokisaari

**Reply by abc-hy** [reply]
For example, I want to measure the characteristic length of these microstructures.

![Picture1](https://github.com/idaholab/moose/assets/63679626/f93da699-4c6b-4840-8187-4ccc82370e55)
![Picture1](https://github.com/idaholab/moose/assets/63679626/66ae2b7f-8cc2-4131-aef7-9d4cf04481df)

**Reply by abc-hy** [reply]
Is there any software can measure the characteristic length of the figure of two phase system?

**Reply by GiudGiud** [reply]
@dschwen

**Comment by laagesen** [comment, marked as answer]
I think you can get it from an analysis of the 2-point correlation function. I talked about using 2-point correlation function for image analysis in this paper and gave a citation for the software we used. I'm not sure off hand exactly how to go from the correlation function to a characteristic length but it might just be the peak of the distribution function (we related it to bubble spacing in our paper and I'm guessing it is the same idea but you will probably need to do a little further reading on this). 

https://www.sciencedirect.com/science/article/pii/S0927025622004839?via%3Dihub

**Reply by abc-hy** [reply]
Thank you for your reply. I read that paper and found that 2 point correlation matlab code. Is this code only suitable for calculating the distance of two circles?

**Reply by laagesen** [reply]
The code calculates the two-point correlation function. To go from the correlation function to characteristic length, you will need to research on your own, I don't know off hand exactly how to do that. However the surface area per unit length I mentioned below is probably easier.

**Comment by laagesen** [comment]
You could also use the inverse of the surface area per unit volume as a length scale. We did this and gave references to where this idea came from in this paper:

https://materialstheory.springeropen.com/articles/10.1186/s41313-021-00041-5

For a 2D simulation, you would use the GB length per unit area. You could get GB length with the GrainBoundaryArea postprocessor (set `grains_per_side = 1`) or with Paraview - in fact I would try both and compare them to make sure you are getting the right answer.

**Reply by abc-hy** [reply]
Thank you for your suggestions. In our work, we did not investigate GB. We only have one order parameter, which is concentration, so this may be not suitable..

**Reply by laagesen** [reply]
The `GrainBoundaryArea` postprocessor should still work for a single order parameter as long as you set `grains_per_side = 1`. Sorry the name is a bit confusing but it should do what you want. You can also use Paraview as I mentioned.

**Reply by abc-hy** [reply]
I am using paraview to output the video of evolution. Could you please tell me how to do this using paraview? Since I don't have grain boundary, which length should I measure?

**Reply by GiudGiud** [reply]
The `GrainBoundaryArea` is a postprocessor in MOOSE, not in paraview.
It wil be a number output to CSV which you can plot in paraview

**Reply by abc-hy** [reply]
![image](https://github.com/idaholab/moose/assets/63679626/999119f0-a5cd-4646-a1a7-dbcb447ccdab)
I added this in postprocessor block, however, I got error message...
![image](https://github.com/idaholab/moose/assets/63679626/4bbd2a48-c764-4741-bda4-a67cab0c192e)
Why is this? Is there other command I need to add?

**Reply by GiudGiud** [reply]
need to add the pf variables

**Reply by TJT-post95** [reply]
According to the source code of `GrainBoundaryArea`, I find that it's the integration of the amplitude about gradient of order parameter within element. Why it's sum could be applied to calculate the length (2D) and area (3D) of grain boundary? Thanks!

**Reply by ChaunceyDevonport** [reply]
I am new to MOOSE and am not well versed enough to look into the source code. I am using two post-processors; the AverageGrainVolume and GrainBoundaryArea post-processors. There are syntax pages in the documentation for the post-processors, but these syntax pages don't explain the calculation that is made for these two post-processors?

**Reply by laagesen** [reply]
AverageGrainVolume uses the FeatureFloodCount (FFC) postprocessor 
https://mooseframework.inl.gov/moose/source/postprocessors/FeatureFloodCount.html
FFC identifies the number of individual features and the volume associated with each. Then AverageGrainVolume totals up those volumes and divides by number of features. (it does not simply divide the total domain volume by number of features to allow it to handle the case where there is a second phase present and you don't want the volume associated with the second phase included in the AverageGrainVolume

**Reply by laagesen** [reply]
GrainBoundaryArea calculates the gradient of each order parameter, then integrates the magnitude of each gradient. You can show analytically that integrating the gradient in 1D will give you the difference between the value of the order parameter on either side of the interface. If you do that over a unit volume and divide by the difference between OPs, you will get the area of the grain boundary (and usually need to divide by a factor of 2 for models where there are 2 order parameters varying through the interface, as is the case the with standard MOOSE grain growth model for polycrystals)

**Reply by TJT-post95** [reply]
OK, thank you!

**Comment by abc-hy** [comment]
What is pf variables?

**Reply by laagesen** [reply]
you need to tell it what phase-field variable you want to use. So if your phase-field order parameter is `c` you need to add `v = c`. I think this should work, I have not actually tried it with the single variable before....if it doesn't work you can also use Paraview.

**Reply by abc-hy** [reply]
Do you mean that I can directly use Paraview to calculate the length? Could you please show me how to realize that?
